habitat-hitl/habitat_hitl/_internal/sim_driver.py

```diff
@@ -0,0 +1,237 @@
+            gui_input=gui_input,
+            remote_client_state=self._remote_client_state,
+            gui_drawer=self._gui_drawer,
+            text_drawer=text_drawer,
+            ui_manager=self._ui_manager,
+            get_anim_fraction=lambda: self._viz_anim_fraction,
+            env=None,
+            sim=self.get_sim(),
+            reconfigure_sim=lambda dataset, scene: self._reconfigure_sim(
+                dataset, scene
+            ),
+            compute_action_and_step_env=None,
+            step_recorder=self._step_recorder,
+            get_metrics=None,
+            end_episode=None,
+            set_cursor_style=self._set_cursor_style,
+            episode_helper=self._episode_helper,
+            client_message_manager=self._client_message_manager,
+            gui_agent_controllers=[],
+            all_agent_controllers=[],
+            users=self._users,
+        )
+
+        self._app_state: AppState = None
+        assert create_app_state_lambda is not None
+        self._app_state = create_app_state_lambda(self._app_service)
+
+        self._reset_environment()
+
+    def _reconfigure_sim(self, dataset: Optional[str], scene: Optional[str]):
+        cfg_settings = habitat_sim.utils.settings.default_sim_settings.copy()
+        # keyword "NONE" initializes a scene with no scene mesh
+        cfg_settings["scene"] = scene if scene else "NONE"
+        cfg_settings["scene_dataset_config_file"] = (
+            dataset if dataset else None
+        )
+        cfg_settings["depth_sensor"] = False
+        cfg_settings["color_sensor"] = False
+        hab_cfg = habitat_sim.utils.settings.make_cfg(cfg_settings)
+        # required for HITL apps
+        hab_cfg.sim_cfg.enable_gfx_replay_save = True
+        self.get_sim().reconfigure(hab_cfg)
+
+        assert self.get_sim().renderer is None
+
+    def _reset_environment(self):
+        if self.network_server_enabled:
+            self._remote_client_state.clear_history()
+
+        self._app_state.on_environment_reset(self._episode_recorder_dict)
+
+        self.get_sim().gfx_replay_manager.save_keyframe()
+
+    def _save_recorded_keyframes_to_file(self):
+        if not self._recording_keyframes:
+            return
+
+        # Consolidate recorded keyframes into a single json string
+        # self._recording_keyframes format:
+        #     List['{"keyframe":{...}', '{"keyframe":{...}',...]
+        # Output format:
+        #     '{"keyframes":[{...},{...},...]}'
+        json_keyframes = ",".join(
+            keyframe[12:-1] for keyframe in self._recording_keyframes
+        )
+        json_content = '{{"keyframes":[{}]}}'.format(json_keyframes)
+
+        # Save keyframes to file
+        filepath = self._save_filepath_base + ".gfx_replay.json.gz"
+        save_as_gzip(json_content.encode("utf-8"), filepath)
+
+    def _set_cursor_style(self, cursor_style):
+        self._pending_cursor_style = cursor_style
+
+    def sim_update(self, dt: float):
```

Comment:
Almost everything here is duplicated in LabDriver.

Reply:
At the moment, I want to avoid adding `habitat_sim` or `habitat_lab` dependencies directly to the base `AppDriver`.

I will do a second refactor pass later.